File: uri/uri_python/grafos/p2440.py
def buscaLargura(G):
	N = len(G)
	cor = []	
	for i in range(N):
		cor.append(-1) #pinta de branco
	k = 1
	floresta = []
	while True:
		x = 0
		while cor[x] != -1:
			x += 1		
		fila = []		
		fila.append(x)
		cor[x] = 0 #pinta de azul
		arvore = []
		while len(fila) > 0:
			x = fila[0]
			fila.remove(fila[0])
			cor[x] = k #pinta de uma cor
			arvore.append(x)
			for i in range(N):
				if G[x][i] and cor[i] == -1:
					cor[i] = 0 #pinta de azul
					fila.append(i)
		floresta.append(arvore)
		flagCor = True
		for i in cor:
			if i == -1:
				flagCor = False
		if flagCor:
			break
		k += 1
	return floresta

# Main
entry = input().split(' ')
V = int(entry[0])
E = int(entry[1])
G = []
for i in range(V):
	G.append([])
	for j in range(V):
		if i == j:
			G[i].append(1)
		else:
			G[i].append(0)
for i in range(E):
	entry = input().split(' ')
	x = int(entry[0]) - 1
	y = int(entry[1]) - 1
	G[x][y] = 1
	G[y][x] = 1
floresta = buscaLargura(G)
print(len(floresta))

File: uri/uri_cpp/grafos/p2440.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <vector>

using namespace std;

int buscaLargura (vector < vector<int> > G);

int main() {
	int i, j, N, V, a, b;
	vector < vector<int> > G;
	cin >> N;
	cin >> V;
	for (i = 0; i < N; i++) {
		vector <int> row;
		for (j = 0; j < N; j++)
			row.push_back(0);		
		G.push_back(row);
	}
	for (i = 0; i < V; i++) {
		cin >> a;
		cin >> b;
		G[a - 1][b - 1] = 1;
		G[b - 1][a - 1] = 1;
	}
	cout << buscaLargura(G) << endl;
	G.clear();
    return 0;
}

int buscaLargura (vector < vector<int> > G) {
	int i, k, x;	
	int N = G.size();
	bool flagCor;
	vector <int> cor;
	for (i = 0; i < N; i++)
		cor.push_back(-1); //pinta de branco
	k = 0;
	while (true) {
		vector <int> fila;
		x = 0;
		while (cor[x] != -1)
			x += 1;
		fila.push_back(x);
		k += 1;
		while (fila.size() > 0) {
			x = fila[0];
			fila.erase(fila.begin()+0);
			cor[x] = k;
			for (i = 0; i < N; i++) {
				if (G[x][i] and cor[i] == -1) {
					cor[i] = 0;
					fila.push_back(i);
				}
			}
		}
		flagCor = true;
		for (i = 0; i < N; i++) {
			if (cor[i] == -1) {
				flagCor = false;
				break;
			}
		}
		if (flagCor)
			break;
		fila.clear();
	}
	cor.clear();
	return k;
}
